Datastore fix for removed taskdef.

hjoliver opened the pull request:
<!--
Thanks for your contribution:
* Please list any related issues with a "closes" or "addresses" tag.
* Please add a helpful description.
* For bugfixes we have maintenance branches e.g. `8.0.x`, please raise separate
  pull requests against master and the maintenance release branches as appropriate.
-->

Close #5057 

Fix datastore bug on restart after removing a task from the graph.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [ ] If this is a bug fix, PRs raised to both master and the relevant maintenance branch.

oliver-sanders commented:
Bumped onto the 8.0.x branch, @hjoliver can you rebase this one.

datamel reviewed (commented):
I've checked this out and run the bug recreation steps from the issue. The traceback appearing on master is fixed, and looking at the gui, the removed task is not present, which is super. 

I have a small query about the logs after restart, removed task b is appearing in the restart logs as failed. Is this by design? Is this confusing for users if their removed task appears in the logs (perhaps apart from maybe being warned that it was once there but now removed). I'm probably missing something!

![image](https://user-images.githubusercontent.com/37735232/185000744-a6ec8272-b7a7-42ef-bfe6-98c8811a351c.png)

dwsutherland commented on cylc/flow/data_store_mgr.py:
```diff
@@ -818,12 +818,15 @@ def generate_ghost_task(self, tp_id, itask, is_parent=False):
 
-        # Most the time the definition node will be in the store,
-        # so use try/except.
+        # Most of the time the definition node will be in the store,
```
Comma at end.

dwsutherland reviewed (approved):
Reproduced, and fix confirmed working.

I suppose we have to filter out DB loads somewhere.

hjoliver commented:
> I have a small query about the logs after restart, removed task b is appearing in the restart logs as failed. 

Good spotting. It's not exactly wrong - if the removed task was still running at shutdown, the scheduler should poll it and log its status and progress (but there would be no further instances of it).  However, I'm not sure why it is being polled here, as it failed before shutdown.  Investigating ...

hjoliver commented:
@datamel  - I've rebased to remove all the spurious commits that appeared after changing the PR base branch (not sure why it did that)... and added a new test.

I'll cherry-pick to master once this is approved.

oliver-sanders commented:
> I'll cherry-pick to master once this is approved.

FYI: we've been merging 8.0.x into master (you can create a PR from the GH web app).

datamel commented on tests/functional/restart/58-removed-task.t:
```diff
@@ -0,0 +1,57 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#-------------------------------------------------------------------------------
+
+GitHub #5067: if a task is removed from the graph after shutdown, it should not
+cause an error at restart. If it was a failed incomplete task, however, it
+should still be polled and logged at restart.
```
```suggestion
# GitHub 5067: if a task is removed from the graph after shutdown, it should not
# cause an error at restart. If it was a failed incomplete task, however, it
# should still be polled and logged at restart.
```

hjoliver replied:
Argghhhhh! Fail!  :cry: :rofl: :face_with_spiral_eyes: :-1: :man_facepalming:  I added the comment at the last minute before pushing the final commit to the branch and leaving work - that'll teach me.

datamel commented on tests/functional/restart/58-removed-task.t:
```diff
@@ -0,0 +1,57 @@
+
+# Incomplete failed tasks a and b should be polled.
+# (Even though b has been removed from the graph).
+grep_workflow_log_ok "grep-3" "\[1/a failed job:01 flows:1\] (polled)failed"
+grep_workflow_log_ok "grep-4" "\[1/b failed job:01 flows:1\] (polled)failed"
+
+# Failed task c should not be polled (not incomplete).
+grep_fail "\[1/c failed job:01 flows:1\] (polled)failed" \
```
This is still failing for me  (with the comments fixed).
I'm not sure if you would rather change the logging or the test to fix?

hjoliver replied:
Hmm, I guess that's potentially flaky, if we can stall and shut down before poll results come in.

hjoliver replied:
(now outdated)

hjoliver commented:
@datamel I've reformulated the functional test to not rely on timing.

hjoliver commented on tests/functional/restart/58-removed-task/flow.cylc:
```diff
@@ -0,0 +1,18 @@
+{% set INCL_B_C = INCL_B_C | default(True) %}
+[scheduler]
+    [[events]]
+        stall timeout = PT0S
```
(now outdated)

datamel reviewed (approved):
All working smoothly for me. Thanks!
